File: test/test_status.py
import pytest

from netft_driver.status import (
    DiagnosticSeverity,
    FtSequenceKind,
    FtSequenceTracker,
    classify_status,
    decode_status,
)


def test_status_severity_distinguishes_health_condition_and_error():
    assert classify_status(0x00000000) is DiagnosticSeverity.OK
    assert classify_status(0x80010000) is DiagnosticSeverity.WARN
    assert classify_status(0x80020000) is DiagnosticSeverity.ERROR
    assert classify_status(0x00010000) is DiagnosticSeverity.ERROR


@pytest.mark.parametrize(
    "mask,name",
    [
        (0x80000000, "error summary"),
        (0x40000000, "CPU or RAM error"),
        (0x20000000, "digital board error"),
        (0x10000000, "analog board error"),
        (0x08000000, "serial link communication error"),
        (0x04000000, "program memory verification error"),
        (0x02000000, "halted due to configuration errors"),
        (0x01000000, "settings validation error"),
        (0x00800000, "configuration incompatible with calibration"),
        (0x00400000, "network communication failure"),
        (0x00200000, "CAN communication error"),
        (0x00100000, "RDT communication error"),
        (0x00080000, "EtherNet/IP protocol failure"),
        (0x00040000, "DeviceNet protocol failure"),
        (0x00020000, "transducer saturation or A/D error"),
        (0x00010000, "monitor condition latched"),
        (0x00004000, "watchdog timeout error"),
        (0x00002000, "stack check error"),
        (0x00001000, "serial EEPROM I2C failure"),
        (0x00000800, "serial flash SPI failure"),
        (0x00000400, "analog board watchdog timeout"),
        (0x00000200, "excessive strain gage excitation current"),
        (0x00000100, "insufficient strain gage excitation current"),
        (0x00000080, "artificial analog ground out of range"),
        (0x00000040, "analog board power supply too high"),
        (0x00000020, "analog board power supply too low"),
        (0x00000010, "serial link data unavailable"),
        (0x00000008, "reference voltage or power monitoring error"),
        (0x00000004, "internal temperature error"),
        (0x00000002, "HTTP protocol failure"),
    ],
)
def test_decode_status_names_every_defined_active_bit(mask, name):
    assert name in decode_status(mask)


def test_decode_status_reports_healthy_when_no_bits_are_set():
    assert decode_status(0) == ("healthy",)


from netft_driver.status import RdtSequenceTracker, SequenceKind


def test_sequence_tracker_classifies_first_contiguous_and_rollover():
    tracker = RdtSequenceTracker()
    assert tracker.observe(0xFFFFFFFE).kind is SequenceKind.FIRST
    assert tracker.observe(0xFFFFFFFF).kind is SequenceKind.CONTIGUOUS
    assert tracker.observe(0).kind is SequenceKind.CONTIGUOUS


def test_sequence_tracker_counts_forward_gap_exactly():
    tracker = RdtSequenceTracker()
    tracker.observe(10)
    observation = tracker.observe(14)
    assert observation.kind is SequenceKind.GAP
    assert observation.gap == 3


def test_sequence_tracker_distinguishes_duplicate_and_out_of_order():
    tracker = RdtSequenceTracker()
    tracker.observe(100)
    assert tracker.observe(100).kind is SequenceKind.DUPLICATE
    assert tracker.observe(99).kind is SequenceKind.OUT_OF_ORDER
    assert tracker.observe(101).kind is SequenceKind.CONTIGUOUS


def test_sequence_tracker_reset_starts_a_new_session():
    tracker = RdtSequenceTracker()
    tracker.observe(50)
    tracker.reset()
    assert tracker.observe(1).kind is SequenceKind.FIRST


def test_ft_sequence_accepts_arbitrary_forward_gaps_and_rollover():
    tracker = FtSequenceTracker()
    assert tracker.observe(0xFFFFFFF0).kind is FtSequenceKind.FIRST
    assert tracker.observe(0xFFFFFFFE).kind is FtSequenceKind.FORWARD
    assert tracker.observe(7).kind is FtSequenceKind.FORWARD


def test_ft_sequence_distinguishes_stall_and_backward_without_moving_baseline():
    tracker = FtSequenceTracker()
    tracker.observe(100)
    assert tracker.observe(100).kind is FtSequenceKind.STALL
    assert tracker.observe(90).kind is FtSequenceKind.BACKWARD
    assert tracker.observe(104).kind is FtSequenceKind.FORWARD


def test_ft_sequence_continues_forward_across_a_new_rdt_session():
    tracker = FtSequenceTracker()
    tracker.observe(1000)
    tracker.begin_session()
    assert tracker.observe(1004).kind is FtSequenceKind.FORWARD


def test_ft_sequence_requires_progress_confirmation_after_new_session_backward():
    tracker = FtSequenceTracker()
    tracker.observe(0x00100000)
    tracker.begin_session()
    assert tracker.observe(2).kind is FtSequenceKind.BACKWARD
    assert tracker.observe(6).kind is FtSequenceKind.RESTART
    assert tracker.observe(10).kind is FtSequenceKind.FORWARD


def test_ft_sequence_confirms_in_session_restart_from_two_low_records():
    tracker = FtSequenceTracker()
    tracker.observe(0x00100000)
    assert tracker.observe(2).kind is FtSequenceKind.BACKWARD
    assert tracker.observe(6).kind is FtSequenceKind.RESTART
    assert tracker.observe(10).kind is FtSequenceKind.FORWARD


def test_ft_restart_candidate_is_cleared_by_session_or_baseline_progress():
    tracker = FtSequenceTracker()
    baseline = 0x00100000
    tracker.observe(baseline)
    assert tracker.observe(2).kind is FtSequenceKind.BACKWARD
    tracker.begin_session()
    assert tracker.observe(6).kind is FtSequenceKind.BACKWARD
    assert tracker.observe(baseline + 4).kind is FtSequenceKind.FORWARD
    assert tracker.observe(10).kind is FtSequenceKind.BACKWARD
    assert tracker.observe(14).kind is FtSequenceKind.RESTART


def test_ft_sequence_does_not_restart_from_progress_just_below_high_baseline():
    tracker = FtSequenceTracker()
    baseline = 0x70000000
    tracker.observe(baseline)
    assert tracker.observe(0x6FFFFF00).kind is FtSequenceKind.BACKWARD
    assert tracker.observe(0x6FFFFF04).kind is FtSequenceKind.BACKWARD
    assert tracker.last == baseline

File: test/cpp/test_status.cpp
#include <gtest/gtest.h>

#include <array>
#include <cstdint>
#include <string>
#include <unordered_set>
#include <utility>

#include "netft_driver/status.hpp"

namespace {
using namespace netft_driver;

TEST(Status, ClassifiesHealthConditionAndError)
{
  EXPECT_EQ(classify_status(0), DiagnosticSeverity::Ok);
  EXPECT_EQ(classify_status(0x80010000U), DiagnosticSeverity::Warn);
  EXPECT_EQ(classify_status(0x80020000U), DiagnosticSeverity::Error);
  EXPECT_EQ(classify_status(0x00010000U), DiagnosticSeverity::Error);
}

TEST(Status, DecodesEveryDefinedActiveBitDistinctly)
{
  constexpr std::array<std::uint32_t, 30> masks{
    0x80000000U, 0x40000000U, 0x20000000U, 0x10000000U, 0x08000000U,
    0x04000000U, 0x02000000U, 0x01000000U, 0x00800000U, 0x00400000U,
    0x00200000U, 0x00100000U, 0x00080000U, 0x00040000U, 0x00020000U,
    0x00010000U, 0x00004000U, 0x00002000U, 0x00001000U, 0x00000800U,
    0x00000400U, 0x00000200U, 0x00000100U, 0x00000080U, 0x00000040U,
    0x00000020U, 0x00000010U, 0x00000008U, 0x00000004U, 0x00000002U};
  std::unordered_set<std::string> descriptions;
  for (const auto mask : masks) {
    const auto description = decode_status(mask);
    EXPECT_FALSE(description.empty());
    descriptions.insert(description);
  }
  EXPECT_EQ(descriptions.size(), masks.size());
}
TEST(Status, ReportsNonemptyDescriptionWhenNoBitsSet) { EXPECT_FALSE(decode_status(0).empty()); }

TEST(RdtSequence, HandlesFirstContiguousRolloverGapsAndOrdering)
{
  RdtSequenceTracker tracker;
  EXPECT_EQ(tracker.observe(0xfffffffeU).kind, SequenceKind::First);
  EXPECT_EQ(tracker.observe(0xffffffffU).kind, SequenceKind::Contiguous);
  EXPECT_EQ(tracker.observe(0).kind, SequenceKind::Contiguous);
  EXPECT_EQ(tracker.observe(4).gap, 3U);
  EXPECT_EQ(tracker.observe(4).kind, SequenceKind::Duplicate);
  EXPECT_EQ(tracker.observe(3).kind, SequenceKind::OutOfOrder);
  EXPECT_EQ(tracker.observe(5).kind, SequenceKind::Contiguous);
  tracker.reset(); EXPECT_EQ(tracker.observe(1).kind, SequenceKind::First);
}

TEST(FtSequence, TracksForwardStallBackwardAndConfirmedRestart)
{
  FtSequenceTracker tracker;
  EXPECT_EQ(tracker.observe(0xfffffff0U).kind, FtSequenceKind::First);
  EXPECT_EQ(tracker.observe(0xfffffffeU).kind, FtSequenceKind::Forward);
  EXPECT_EQ(tracker.observe(7).kind, FtSequenceKind::Forward);
  EXPECT_EQ(tracker.observe(7).kind, FtSequenceKind::Stall);
  EXPECT_EQ(tracker.observe(6).kind, FtSequenceKind::Backward);
  tracker.observe(0x00100000U); tracker.begin_session();
  EXPECT_EQ(tracker.observe(2).kind, FtSequenceKind::Backward);
  EXPECT_EQ(tracker.observe(6).kind, FtSequenceKind::Restart);
  EXPECT_EQ(tracker.observe(10).kind, FtSequenceKind::Forward);
}

TEST(FtSequence, ClearsCandidatesAndRejectsHighWindowNearProgress)
{
  FtSequenceTracker tracker; const auto baseline = 0x00100000U;
  tracker.observe(baseline); EXPECT_EQ(tracker.observe(2).kind, FtSequenceKind::Backward);
  tracker.begin_session(); EXPECT_EQ(tracker.observe(6).kind, FtSequenceKind::Backward);
  EXPECT_EQ(tracker.observe(baseline + 4).kind, FtSequenceKind::Forward);
  EXPECT_EQ(tracker.observe(10).kind, FtSequenceKind::Backward);
  EXPECT_EQ(tracker.observe(14).kind, FtSequenceKind::Restart);
  FtSequenceTracker high; high.observe(0x70000000U);
  EXPECT_EQ(high.observe(0x6fffff00U).kind, FtSequenceKind::Backward);
  EXPECT_EQ(high.observe(0x6fffff04U).kind, FtSequenceKind::Backward);
  EXPECT_EQ(high.last(), 0x70000000U);
}

}  // namespace
